keps/sig-storage/20200212-volume-group.md

```diff
@@ -0,0 +1,945 @@
+
+## Proposal for Consistency Groups and Group Snapshots
+
+This proposal introduces new CRDs VolumeGroup, VolumeGroupClass, and GroupSnapshot.
+
+Create new VolumeGroup can be done in several ways:
+1. Create an empty group first, then create a new PVC with group_id which will add a volume to the already created group.
+2. Create an empty group first, and then add an existing PVC to the group one by one.
+3. Non-goal: Immutable volume group will not be supported. Immutable volume group means a volume's membership in a volume group cannot be changed once the group is created. This means the following way of creating a group is not supported - creating a group with existing volumes, either with a list of PVCs or using a selector.
+4. Non-goal: Create a new empty group and in the same time create new PVCs and add to the new group.
+
+Modify an existing VolumeGroup:
+Add new volume or remove existing volume from an existing VolumeGroup. Option 2 for create VolumeGroup above falls into this case.
+
+Two types of VolumeGroups can be supported:
+
+Type 1: Mutable VolumeGroup - Create new PVC and add to the group.
+* Create a new empty VolumeGroup.
+* Create a new PVC with existing VolumeGroup name. As a result, new PVC is created and added to VolumeGroup. VolumeGroup is modified so Status has this new PVC in PVCList.
+* External-provisioner will be modified so that VolumeGroupName will be passed to the CSI driver when creating a volume.
+
+Only CSI drivers supporting VOLUMEGROUP capability can support this type of mutable volume group.
+When a new PVC is created with the existing VolumeGroup name, the VolumeGroup will be modified and the PVC will be added to PVCList in the Status.
+
+Type 2: Mutable VolumeGroup - Add existing PVC to the group or remove PVC from the group without deleting it. A VOLUMEGROUP_ADD_REMOVE_EXISTING_VOLUME capability will be added to CSI Spec. Only CSI drivers supporting both VOLUMEGROUP and VOLUMEGROUP_ADD_REMOVE_EXISTING_VOLUME capabilities can support this type of mutable volume group.
+* Type 2 is a subset of Type 1 VolumeGroup.
+* Create a new empty VolumeGroup. VolumeGroup Status has a boolean AddRemoveExistingPVC set to true.
+* Add an existing PVC to an existing VolumeGroup (VolumeGroup can be empty to start with or it can have other PVCs already) by adding VolumeGroup name to the PVC Spec.
+* VolumeGroup is modified so the existing PVC is added to the PVCList in the Status.
+* External-provisioner will be modified so that modifying PVC by adding VolumeGroupName will trigger a ModifyVolume call (a new CSI controller RPC) to CSI driver.
+* VolumeGroup controller will be triggered to update the VolumeGroup Status.
+* Deleting a PVC from a VolumeGroup will trigger external-provisioner and the VolumeGroup controller as well.
+
+CSI drivers that do not have a volume_group_id concept can use the VolumeGroup name stored in Kubernetes API server as the volume_group_id.
+
+Snapshot of a VolumeGroup:
```

Comment:
A storage system MAY create a single snapshot, referred to by a single snapshot-id, for all volumes in the group for a CreateGroupSnapshot request. In such cases what is the intended response to a DeleteSnapshot request, for a volume that belongs to a group and further has a GroupSnapshotName in the VolumeSnapshot object being deleted?
- Should the DeleteSnapshot delete the group snapshot?
  - NOTE: Considering the request is targeted at a single VolumeSnapshot, and the storage system is creating a single snapshot for all volumes in the group, this is not an expected outcome
- Will such a DeleteSnapshot (for a VolumeSnapshot that contains GroupSnapshotName) be disallowed, and the only manner of deletion would be via the DeleteGroupSnapshot?

From the perspective of the storage driver, it can return appropriate errors for a DeleteSnapshot in such cases and only allow the delete when the entire GroupSnapshot is deleted, but asking anyway to clarify thoughts around the same.

Reply:
DeleteSnapshot on a single snapshot that belongs to a group snapshot is not allowed.